langserver/handler.go

```diff
@@ -71,19 +67,24 @@ func (h *LangHandler) resetCaches(lock bool) {
 	if lock {
 		h.mu.Lock()
 	}
-	h.cache = map[typecheckKey]*typecheckResult{}
+
 	h.importGraphOnce = sync.Once{}
 	h.importGraph = nil
-	if lock {
-		h.mu.Unlock()
+
+	if h.typecheckCache == nil {
+		h.typecheckCache = newTypecheckCache()
+	} else {
+		h.typecheckCache.Purge()
```

Comment:
Hmm, but this is a bit weird that `h.typecheckCache == nil` determines if `h.typecheckCache` can have a race condition or not. I'll approve this PR now, I trust you to keep this in mind. :-)